src/resolve/resolved-manager.c

```diff
@@ -1054,6 +1054,58 @@ static int manager_ipv6_send(
         return sendmsg_loop(fd, &mh, 0);
 }
 
+int send_dns_notification(Manager *m, DnsAnswer *answer, const char *query_name) {
+        _cleanup_free_ char *normalized = NULL;
+        DnsResourceRecord *rr;
+        int ifindex, r;
+        _cleanup_(json_variant_unrefp) JsonVariant *array = NULL;
+        Varlink *connection;
+
+        assert(m);
+
+        DNS_ANSWER_FOREACH_IFINDEX(rr, ifindex, answer) {
+                _cleanup_(json_variant_unrefp) JsonVariant *entry = NULL;
+
+                if (rr->key->type == DNS_TYPE_A) {
+                        struct in_addr *addr = &rr->a.in_addr;
+                        r = json_build(&entry,
+                                JSON_BUILD_OBJECT(JSON_BUILD_PAIR_CONDITION(ifindex > 0, "ifindex", JSON_BUILD_INTEGER(ifindex)),
+                                                  JSON_BUILD_PAIR("family", JSON_BUILD_INTEGER(AF_INET)),
+                                                  JSON_BUILD_PAIR_IN4_ADDR("address", addr),
+                                                  JSON_BUILD_PAIR("type", JSON_BUILD_STRING("A"))));
+                } else if (rr->key->type == DNS_TYPE_AAAA) {
+                        struct in6_addr *addr6 = &rr->aaaa.in6_addr;
+                        r = json_build(&entry,
+                                JSON_BUILD_OBJECT(JSON_BUILD_PAIR_CONDITION(ifindex > 0, "ifindex", JSON_BUILD_INTEGER(ifindex)),
+                                                  JSON_BUILD_PAIR("family", JSON_BUILD_INTEGER(AF_INET6)),
+                                                  JSON_BUILD_PAIR_IN6_ADDR("address", addr6),
+                                                  JSON_BUILD_PAIR("type", JSON_BUILD_STRING("AAAA"))));
+                } else
+                        continue;
```

Comment:
missing error handling.

BTW, if you'd like, you can also use `JSON_BUILD_PAIR_INTEGER` and `JSON_BUILD_PAIR_STRING`.

Reply:
Added the error handling (log message and continue) and used the new macros. Thanks for the suggestion!